api: Implement common protocol for RabbitMQ messages

victorges opened the pull request:
<!-------------------------------------------------------------------------
 | Thanks for send a pull request! 🎉
 | First, please make sure you familiar with the contribution guidelines
 | https://github.com/livepeer/livepeer.com/blob/master/CONTRIBUTING.md
 -------------------------------------------------------------------------->

**What does this pull request do? Explain your changes. (required)**
This is to implement the common protocol proposed with the `data/analyzer` logic for 
events sent from our services into RabbitMQ. The main reason for this protocol is that
we need all events to be serialized on the **same** data stream (~a queue), while still being
able to process them correctly according to their corresponding type. In the low-level, this
meant:
 - a `type` field which specifies what is the type of the message. This field should be parsed
 first to determine how to parse the rest of the event (in [data/analyzer](https://github.com/livepeer/livepeer-data/blob/5b75a623cbb307520e42001663532c1572f6b46b/pkg/data/parse.go#L13))
 - a `timestamp` field instead of a `createdAt`, which is how I had followed for `data/analyzer`
 and I think it makes sense for events/messages. 
 - a required `id` field because why not, should be helpful
 - a required `manifestId`, since not all Livepeer services understand stream IDs so that's the global stream identifier

2 things here that I intend to update `livepeer-data` for consistency (and functionality):
 - Timestamps should be in milliseconds instead of nanoseconds. Nano required `int64` and that's just
 not portable. JSON is not defined that way, and node doesn't even know how to serialize `bigint`s. This
 is a backward-incompatible change on `analyzer`, but I intend to do anyway since we're only in `staging` yet.
 - Will rename "events" to "messages" instead, which is what they actually are. In the API we already
 had the higher-level user-facing "events", so we should better keep those separate. This is just nomenclature
 in the code, so lower-pri and should have no incompatibilities.

**Specific updates (required)**
 - Improve data stored in `webhook_response` table, for more debuggability
 - Create separate types for the 2 cannon RabbitMQ messages (event and trigger)
 - Fix webhook retries (I think either the N*N bug was back or retries were broken)
 - Implement that common message protocol described above
 - Move `sessionId` field of `recording.ready` event out of the user-facing `payload`

## -

- **How did you test each of these updates (required)**
`yarn test` 
Mostly some structural changes on code very well exercised in the tests, so I believe
this is enough. Also added a test for webhook retries to make sure they keep working. 

**Does this pull request close any open issues?**
This is part of the movement to get `multistream`s in prod, with connectivity data available
in stream health/`data/anlyzer`.

**Checklist:**

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the **CONTRIBUTING** document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

victorges committed:
api: Improve resp info on webhook_response

We were not serializing much about the repsonse
other than the statusCode, since the native
`Response` object from fetch doesn't serialize
nicely as JSON. So we have to extract some of
the properties that we want from it.

victorges committed:
api: Also store how long it took to call a webhook

victorges committed:
api/webhooks: Create separate message types ns/file

victorges committed:
api/webhooks: Create separate WebhookTrigger msg

This also fixes retries, which actually seemed to be broken.
They sent an "event" is if it were a trigger for retry, and
most probably err'd on the second try.

victorges committed:
api/store: Add some type safety to rabbit-queue

Now we it will be easier to guarantee we follow the
global RabbitMQ protcol being defined between livepeer
infra services.

Messages should always have a `type` key specifying what's
the type of the object so it can be deserialized properly.
Also made sure webhook event routing keys always include
the event itself, so one can consistently grab all stream,
multistream, recording events, etc.

victorges committed:
api/webhooks: Only ack/nack webhook event once

With the try/catch block inside the triggers.map
transformation, it meant we could nack the event
multiple times, once for each webhook that failed.

Additionally, the `return` in there didn't mean much,
since we would only return from the inner lambda but
not from the outer handleEventsQueue function.

Moved try/catch to outside the Promise.all, so that
we only nack once and make sure to return from
handleEventsQueue and avoid the `ack` in the end of
the function.

victorges committed:
api/store: Implement RabbitMQ livepeer protocol

victorges committed:
api/webhooks: Move sessionId field out of payload

Since the information is not meant to the user but for our
own cannon logic, we can make it a first party field in the
RabbitMQ message. IMO ideally, the payload should contain
only information that should be sent on the webhook call,
not for our internal use.

darkdarkdragon commented on packages/api/src/controllers/stream.ts:
```diff
@@ -888,11 +893,13 @@ app.put(
-      channel: "webhooks",
-      event: req.body.active === true ? "stream.started" : "stream.idle",
+      timestamp: Date.now(),
+      manifestId: stream.playbackId,
```
nowadays, in our codebase, `manifestId` can mean many different things. 
If we're passing `playbackId` - maybe it's better to call it `playbackId`? That (still 😄 ) has only one meaning 😄

victorges replied:
Ideally, here we would use some kind of "global identifier" which is known by all Livepeer services and uniquely identifies each stream in our infrastructure. So that a service processing all the messages on RabbitMQ would be able to join all events from a given stream using only the ID value in that respective field.

Since `go-livepeer` apparently has no reference to `playbackId`s ([only one I found was a comment for a field that is actually `ManifestID`](https://github.com/livepeer/go-livepeer/blob/72047f1d2d95b89a308da1064f3f3be08c89d1ec/core/streamdata.go#L165)), I __think__ it makes more sense to stick with `manifestId` for the API-agnostic field here, unless you have some other idea. Some alternatives that came to my mind but are more like questions:
 - Could we assume on `go-livepeer` that every `manifestId = playbackId` and set a field called `playbackId` with the value known there as `manifestId` instead? So in the events we have `playbackId`, and on `go-livepeer` we set `event.playbackId = mid`, instead of the opposite here
 - Do we have a way to determine the "true" `manifestId` here? That since it can be set to different things, not only the `playbackId`, is that right? Or is it truly non-deterministic for the same stream, where different services use different values but call all of them `manifestId`?

I guess the more open questiion is: What's the best way to create this "global identifier" to be used by a service joining all events about a given stream?

victorges replied:
Resolved by using the `streamId` as the global stream identifier.

 - livepeer-data: https://github.com/livepeer/livepeer-data/pull/7/commits/029683482b09a135f5b13a620213b43ec6dfb55c
 - go-livepeer: https://github.com/livepeer/go-livepeer/pull/2002/commits/b814dfb08208f1857b7bb6e01ac448246b006140
 - livepeer-com: https://github.com/livepeer/livepeer-com/pull/569/commits/74f1a4cb2a055be10fbf29b1db87bdb74ab7f02a
 - stream-tester: https://github.com/livepeer/stream-tester/pull/96/commits/b2ee197b342a7dc4499feeea8e1e9acbb59fa9a1

darkdarkdragon commented on packages/api/src/webhooks/cannon.ts:
```diff
@@ -144,29 +144,21 @@ export default class WebhookCannon {
     }
     try {
       // TODO Activate URL Verification
-      if (message.event && message.webhook && message.stream && message.user) {
```
Why you've removed this check?

victorges replied:
Those are already checked in the fist line of the `_fireHook` function: https://github.com/livepeer/livepeer-com/blob/ebb97569b815880295da59fb530bb8aa88e550ff/packages/api/src/webhooks/cannon.ts#L218-L223

It also throws an error, which I think is better since it will be logged in the catch below. Just realized that the message will be unwillingly retried on that case, so I can fix that if you prefer. But IMO this is more of a sanity check that should never happen in production, so we don't want to silently ignore those cases anyway. 

I guess that we only really needed it before because of that `retry` bug that was actually sending the original `event` to the `*_single_url` queue, instead of the actual "webhook trigger" as I'm calling it here. With that our retries were actually broken, only 1 attempt was made per webhook. This new test I created won't let that happen again :)

darkdarkdragon replied:
> Just realized that the message will be unwillingly retried on that case, so I can fix that if you prefer

Yes, and that error will send message to retry loop. 
I've hit this problem when was developing and my old code sent message to the new one.

Everywhere we're getting data from external system we should check that is has format we expect. 
I think in this case we safely can just ignore such message.

victorges replied:
Fixed: https://github.com/livepeer/livepeer-com/pull/569/commits/340eb2dc808943f69470d4ae839b4e6247aceaef

victorges committed:
api/webhooks: Fix tests (among other things)

Also fix the new event queue, to avoid having the
same binding as before getting all events in the
default queue.

Also fix delayed messages emitting. We were
both accumulating duplicate setup functions as
well as re-using the same queues for different
routing keys.

victorges committed:
api/webhooks: Create test for webhook retries

victorges committed:
api/webhooks: Create helper processWebhookEvent

I broke the tests initially because I forgot to ack
the message. Since it seems very error prone to require
individual ack/nacks like that, created a helper just
like for trigger processing, ack/nacking the message on
the finally block.

victorges committed:
Fix tests

forgot a final return on the helper created on previous commit

iameli commented on packages/api/src/webhooks/cannon.test.ts:
```diff
@@ -219,29 +228,34 @@ describe("webhook cannon", () => {
 
-      const sem = semaphore();
-      let callCount = 0;
+      const semaphs = [semaphore(), semaphore()];
```
`sems` not good enough for you? 😁

victorges replied:
It actually sounded a bit cacophonic to me, but no biggie. Will change to `sems` 😁

victorges committed:
api/webhooks: Log bad event instead of throwing error

This will avoid the infinite retry loop on a bad event,
but still avoid keeping us blind in case bad events start
lurking into the internal queue. That must never happen.

victorges committed:
s/semaphs/sems/g

victorges committed:
api/stream: manifestID=>streamID on rabbit events

darkdarkdragon reviewed (approved):
LGTM
